ADApp/ntndArrayConverterSrc/ntndArrayConverter_pvxs.cpp

```
@@ -0,0 +1,447 @@
+    m_fieldNameMap = {
+        {typeid(int8_t), "value->byteValue"},
+        {typeid(uint8_t), "value->ubyteValue"},
+        {typeid(int16_t), "value->shortValue"},
+        {typeid(uint16_t), "value->ushortValue"},
+        {typeid(int32_t), "value->intValue"},
+        {typeid(uint32_t), "value->uintValue"},
+        {typeid(int64_t), "value->longValue"},
+        {typeid(uint64_t), "value->ulongValue"},
+        {typeid(float_t), "value->floatValue"},
+        {typeid(double_t), "value->doubleValue"}
+    };
+
+    m_arrayTypeMap = {
+        {typeid(int8_t), pvxs::ArrayType::Int8},
+        {typeid(uint8_t), pvxs::ArrayType::UInt8},
+        {typeid(int16_t), pvxs::ArrayType::Int16},
+        {typeid(int32_t), pvxs::ArrayType::Int32},
+        {typeid(uint32_t), pvxs::ArrayType::UInt32},
+        {typeid(int64_t), pvxs::ArrayType::Int64},
+        {typeid(uint64_t), pvxs::ArrayType::UInt64},
+        {typeid(float_t), pvxs::ArrayType::Float32},
+        {typeid(double_t), pvxs::ArrayType::Float64}
+    };
+
+}
+
+NDColorMode_t NTNDArrayConverter::getColorMode (void)
+{
+    auto attributes = m_value["attribute"].as<pvxs::shared_array<const pvxs::Value>>();
+    NDColorMode_t colorMode = NDColorMode_t::NDColorModeMono;
+    for (int i=0; i<attributes.size(); i++) {
+        pvxs::Value attribute = attributes[i];
+        if (attribute["name"].as<std::string>() == "ColorMode") {
+            colorMode = (NDColorMode_t) attribute["value"].as<int32_t>();
+            break;
+        }
+    }
+    return colorMode;
+}
+
+NTNDArrayInfo_t NTNDArrayConverter::getInfo (void)
+{
+    NTNDArrayInfo_t info = {0};
+
+    auto dims = m_value["dimension"].as<pvxs::shared_array<const pvxs::Value>>();
+    info.ndims = (int) dims.size();
+    info.nElements = 1;
+
+    for(int i = 0; i < info.ndims; ++i)
+    {
+        info.dims[i]    = dims[i]["size"].as<size_t>();
+        info.nElements *= info.dims[i];
+    }
+
+    info.codec = m_value["codec.name"].as<std::string>();
```

More relevant with pvaDriver, so fyi. here.  Be aware that `as<>()` will throw an exception if the `Value` is empty.  So this line will fail hard if `codec` is missing.

In situations where the structure definition is not provided by local code, eg. client code working with a server provided Value, I would recommend a more relaxed test of any field which are not essential.

There are two overloads of `Value::as()` to assist with this:

```c++
// if info.codec is already std::string
(void)m_value["codec.name"].as<std::string>(info.codec); // returns true if info.codec is updated

m_value["codec.name"].as<std::string>([&info](const std::string& v) {
    info.codec = v;
});
```

Or alternately

```c++
if(auto name = m_value["codec.name"])
    info.codec = name.as<std::string>();
```

In the first two cases both absence and type mis-match are handled softly.  In the last, only absense.